lib/ngx/encoding.lua

```diff
@@ -0,0 +1,76 @@
+-- Copyright (C) by Yichun Zhang (agentzh)
+-- Copyright (C) by OpenResty Inc.
+
+-- I hereby assign copyright in this code to the lua-resty-core project,
+-- to be licensed under the same terms as the rest of the code.
+--
+local ffi = require("ffi")
+local base = require("resty.core.base")
+local ffi_str = ffi.string
+local C = ffi.C
+local NGX_ERROR = ngx.ERROR
+
+
+local _M = { version = base.version }
+
+
+ffi.cdef[[
+typedef intptr_t        ngx_int_t;
+
+void ngx_encode_base64url(ngx_str_t *dst, ngx_str_t *src);
+ngx_int_t ngx_decode_base64url(ngx_str_t *dst, ngx_str_t *src);
+]]
+
+
+local get_string_buf = base.get_string_buf
+
+
+local dst = ffi.new("ngx_str_t[1]")
+local src = ffi.new("ngx_str_t[1]")
+
+local function base64_encoded_length(len)
+    return ((len + 2) / 3) * 4
+end
+
+
+local function base64_decoded_length(len)
+    return ((len + 3) / 4) * 3
+end
+
+
+function _M.encode_base64url(s)
+    local len = #s
+    local trans_len = base64_encoded_length(len)
+
+    src[0].data = s
+    src[0].len = len
```

Comment:
Better cache the value `src[0]` in a local variable since it is referenced for more than one time.